### PR TITLE
Remove DependencyResolverInterface

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌

### CODE REVIEW

File: src/ArrayDefinition.php [hunk 1]
```diff
@@ -32,6 +32,7 @@ final class ArrayDefinition implements DefinitionInterface
      */
     private string $class;
     private array $constructorArguments;
+    private ?ContainerInterface $referenceContainer = null;
```
Comment (samdark) [inline src/ArrayDefinition.php:35]:
```suggestion
    /**
      * Container used to resolve references.
      */
    private ?ContainerInterface $referenceContainer = null;
```

**Comment by samdark** [comment]
https://github.com/yiisoft/db/pull/237

File: src/Infrastructure/DefinitionResolver.php [hunk 2]
```diff
@@ -58,14 +58,15 @@ public static function resolveArray(DependencyResolverInterface $dependencyResol
...
      * @return mixed
      */
-    public static function resolve(DependencyResolverInterface $dependencyResolver, $definition)
+    public static function resolve(ContainerInterface $container, ?ContainerInterface $referenceContainer, $definition)
     {
         if ($definition instanceof DefinitionInterface) {
+            $container = $referenceContainer !== null && $definition instanceof ReferenceInterface ? $referenceContainer : $container;
             /** @var mixed $definition */
-            $definition = $definition->resolve($dependencyResolver);
+            $definition = $definition->resolve($container);
```
Comment (vjik) [inline src/Infrastructure/DefinitionResolver.php:66]: Don't good override exists variable `$container`. Suggestion:

```php
$definition = $definition->resolve(
	$referenceContainer !== null && $definition instanceof ReferenceInterface
		? $referenceContainer
		: $container
);
```

File: tests/Unit/Infrastructure/DefinitionExtractorTest.php [hunk 3]
```diff
@@ -49,22 +49,22 @@ public function testResolveConstructor(): void
...
         /** @var DefinitionInterface[] $dependencies */
         $dependencies = $resolver->fromClassName(Car::class);
 
         $this->assertCount(2, $dependencies);
         $this->assertInstanceOf(ClassDefinition::class, $dependencies['engine']);
         $this->assertInstanceOf(ParameterDefinition::class, $dependencies['moreEngines']);
 
-        $this->expectException(NotFoundException::class);
-        $dependencies['engine']->resolve($dependencyResolver);
+        $this->expectException('Yiisoft\Test\Support\Container\Exception\NotFoundException');
```
Comment (vjik) [inline tests/Unit/Infrastructure/DefinitionExtractorTest.php:60]:
```suggestion
        $this->expectException(NotFoundException::class);
```

Comment (yiiliveext) [reply tests/Unit/Infrastructure/DefinitionExtractorTest.php:60]: You are wrong, we have two different  `NotFoundException` classes in this test.

File: src/Infrastructure/DefinitionResolver.php (hunk 2 above)
Comment (yiiliveext) [reply src/Infrastructure/DefinitionResolver.php:66]: Why not?

File: tests/Unit/Infrastructure/DefinitionExtractorTest.php (hunk 3 above)
Comment (vjik) [reply tests/Unit/Infrastructure/DefinitionExtractorTest.php:60]: Use full name `\Yiisoft\Test\Support\Container\Exception\NotFoundException::class`. Is better that string.

File: src/Infrastructure/DefinitionResolver.php (hunk 2 above)
Comment (vjik) [reply src/Infrastructure/DefinitionResolver.php:66]: bad practic

**Pull request merged**